src/Resources/contao/library/Contao/Controller.php

```diff
@@ -1113,8 +1104,20 @@ public static function generateFrontendUrl(array $arrRow, $strParams=null, $strF
 			$arrParams['_locale'] = $objPage->rootLanguage;
 		}
 
-		$strUrl = $objRouter->generate($strRoute, $arrParams);
-		$strUrl = substr($strUrl, strlen(\Environment::get('path')) + 1);
+		// Add the domain if it differs from the current one (see #3765 and #6927)
+		if ($blnFixDomain)
+		{
+			$arrParams['_domain'] = $arrRow['domain'];
+			$arrParams['_ssl'] = (bool) $arrRow['rootUseSSL'];
+		}
+
+		$strUrl = $objUrlGenerator->generate($strAlias, $arrParams);
+
+		// Remove path from absolute URLs
```

Comment:
We don't want `RELATIVE_PATH`, it would return something like `../../foobar.html` (depending on your current page).